### DISCUSSION TITLE
How to implement isOnline

### DISCUSSION THREAD

**Discussion by jeffalo** [ScratchAddons/ScratchAddons#504] [Ideas]
ok so w_l has told me that having a central server has many issues, which is ok but in my opinion putting random invisible characters into people's bios doesn't seem to be that good either.

i have come to the ultimate solution!!!!

so, the idea is, we use scratch's own cloud servers to do it.

"but jeffalo, thats spamming scratch's servers", i hear you asking. fear not!! we only need to update a variable every couple minutes and plus anyways most cloud projects update the variables many times a second! (probably)

we can spread the updating of cloud across multiple projects to maximize efficiency or something like that.

by piggybacking on scratch's servers we can avoid having to figure out if a user is actually the right person, because scratch does that for us.

then, we'd use a central server (don't worry technically we don't need to <sub><sup>ok but realistically we do</sup></sub>) which watches the cloud logs and sees whos changing it, then they'll be added to an array and considered online for x minutes (probably like 4 or something) and then they are removed

over on scratch. when we visit someone's profile after loading the page we make a fetch to the server asking "yo is person online?" and then the server says if they're online or not.

someone is considered online if they're in the array, and if they're not then they're offline. when someone is offline we don't display anything (because they might not be using the addon / extension etc etc) but if they're online, then we display a little green dot or something.

i can write the server code and addon probably :)

what do you think?

# tldr;

## run in background
step 1 update cloud variable on client every few miniutes
step 2 read cloud updates using server, we dont need to chekc if they're authorized because scratch does it for us
step 3 add them to an array and remove them in x miniutes

## when user gos to profile 
step 1: when page loads, fetch some api on the server asking if the profile's visited user is online
step 2: the server returns weather the user is in the array
step 3: if the server returns true, then show the green dot, otherwise do nothing
![image](https://user-images.githubusercontent.com/40470736/96263679-8fc4a280-0fc3-11eb-8ccb-ca4921895516.png)

## when user is in cloudgames menu(maybe this is bad idea?)
step 1: fetch a list of online users from the server
step 2: server returns a list of users in the array
step 3: list is rendered in the cloud games menu

we can also consider users who are online in cloud games as online on scratch. this way they don't even need the addon to be checked if online.

# tldr tldr;
basically what the cloud games menu does right now, but on a larger scale, displays on profiles and probably with a central server to speed things up

**Comment by Hans5958** [comment]
Bruh, we are going back into square one that is problematic, which was using the Scratch services (which was the country).

In the other hand, should I move #342 to here, the Discussions forum?

**Reply by jeffalo** [reply]
that issue is old and doesn't really explain it as well as this one imo.

we're not going to be using scratch services, _that hard_, i'm sure updating peoples' bios would be much more intense for scratch _cough @WorldLanguages_ 

but i understand if this might be a bit bad if we just piggyback scratch's stuff...

**Reply by GrahamSH-LLK** [reply]
Yeah. I feel like piggybacking isn't a great idea.

**Reply by jeffalo** [reply]
> Yeah. I feel like piggybacking isn't a great idea.

yeah, thats true, although there wouldn't be _that_ many cloud changes, and plus cloud was designed to be updated quickly...

**Comment by jeffalo** [comment]
i made a really really simple online indicator serverbased on the cloud multiplayer online list code.

https://repl.it/@jeffalobob/online#index.js and https://scratch.mit.edu/projects/437219153/

**Reply by GrahamSH-LLK** [reply]
Edit: I get it! That's a great idea! All that we need now is a way to send to the clouddata server. What about new scratchers?

**Reply by jeffalo** [reply]
thanks! well its kinda meh about newscratchers, there probably aren't that many new scratchers using scratchaddons, and we could just say that they need to be scratchers to use this addon i guess.

sending data to cloud server can't be that hard but i have no clue how..

also i saw you online!
![image](https://user-images.githubusercontent.com/40470736/96270185-9a833580-0fcb-11eb-90e6-568d0a76dc51.png)

**Reply by GrahamSH-LLK** [reply]
Ok. One thing I noticed is that the server seems to go offline every so often. Is that you editing it?

**Reply by jeffalo** [reply]
> Ok. One thing I noticed is that the server seems to go offline every so often. Is that you editing it?

yeah thats just repl.it, i'm editing it

**Comment by retronbv** [comment]
So cool!

**Comment by Explosion-Scratch** [comment]
Cool!

**Comment by FunctionalMetatable** [comment]
will this be implemented?
